pkg/daemon/update.go

```diff
@@ -1650,50 +1649,89 @@ func (dn *Daemon) updateSSHKeys(newUsers []ign3types.PasswdUser) error {
-				if err != nil {
-					return fmt.Errorf("failed to remove path '%s': %w", authKeyPath, err)
-				}
-			} else if !os.IsNotExist(err) {
-				// This shouldn't ever happen
-				return fmt.Errorf("unexpectedly failed to get info for path '%s': %w", authKeyPath, err)
-			}
 
-			// Ensure authorized_keys.d/ignition is the only fragment that exists
-			keyFragmentsDir, err := ctrlcommon.ReadDir(authKeyFragmentDirPath)
-			if err == nil {
-				for _, fragment := range keyFragmentsDir {
-					if fragment.Name() != "ignition" {
-						keyPath := filepath.Join(authKeyFragmentDirPath, fragment.Name())
-						err := os.RemoveAll(keyPath)
-						if err != nil {
-							return fmt.Errorf("failed to remove path '%s': %w", keyPath, err)
-						}
-					}
-				}
-			} else if !os.IsNotExist(err) {
-				// This shouldn't ever happen
-				return fmt.Errorf("unexpectedly failed to get info for path '%s': %w", authKeyFragmentDirPath, err)
+	authKeyPath := constants.RHCOS8SSHKeyPath
+
+	if !dn.mock {
+		// In RHCOS 8.6 or lower, the keys were written to `/home/core/.ssh/authorized_keys`.
+		// RHCOS 9.0+, FCOS, and SCOS will however expect the keys at `/home/core/.ssh/authorized_keys.d/ignition`.
+		// Check if the authorized_keys file at the legacy path exists. If it does, remove it.
+		// It will be recreated at the new fragment path by the atomicallyWriteSSHKey function
+		// that is called right after.
+		if dn.useNewSSHKeyPath() {
+			authKeyPath = constants.RHCOS9SSHKeyPath
+
+			if err := cleanSSHKeyPaths(); err != nil {
+				return err
 			}
 		}
 
 		// Note we write keys only for the core user and so this ignores the user list
-		if err := dn.atomicallyWriteSSHKey(concatSSHKeys); err != nil {
-			return err
+		return dn.atomicallyWriteSSHKey(authKeyPath, concatSSHKeys)
+	}
+
+	return nil
+}
+
+// Determines if a file exists by checking for the presence or lack thereof of
+// an error when stat'ing the file. Returns any other error.
+func fileExists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	// If there is no error, the file definitely exists.
+	if err == nil {
+		return true, nil
+	}
+
+	// If the error matches fs.ErrNotExist, the file definitely does not exist.
+	if errors.Is(err, fs.ErrNotExist) {
+		return false, nil
+	}
+
+	// An unexpected error occurred.
+	return false, fmt.Errorf("cannot stat file: %w", err)
+}
+
+// Removes the old SSH key path (/home/core/.ssh/authorized_keys), if found.
+func cleanSSHKeyPaths() error {
+	oldKeyExists, err := fileExists(constants.RHCOS8SSHKeyPath)
+	if err != nil {
+		return err
+	}
+
+	if !oldKeyExists {
+		return nil
+	}
+
+	if err := os.RemoveAll(constants.RHCOS8SSHKeyPath); err != nil {
+		return fmt.Errorf("failed to remove path '%s': %w", constants.RHCOS8SSHKeyPath, err)
+	}
+
+	return removeNonIgnitionKeyPathFragments()
+}
+
+// Ensures authorized_keys.d/ignition is the only fragment that exists within the /home/core/.ssh dir.
+func removeNonIgnitionKeyPathFragments() error {
```

Comment:
Why do we need to remove remained files in authorized_keys.d/ ? Someone can ideally add it using regular ignition. Is this an undesired behavior on RHCOS?

Reply:
I'm not sure. This behavior was introduced in https://github.com/openshift/machine-config-operator/commit/9bae8d623acc5944630cbea14bc7e94e95ed0d93. I just refactored it into a separate function for readability.